temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java

```diff
@@ -290,17 +314,23 @@ public void close() {
       lock.unlock();
       return;
     }
-    // Do not close while runUntilAllBlocked executes.
-    // closeRequested tells it to call close() at the end.
     closeRequested = true;
-    if (inRunUntilAllBlocked) {
+    if (
```

Comment:
The main important change in this PR is here, everything else are refactorings, cleanups, and comments.